### PR TITLE
RGO2018

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
PLEASE CHECK IF THIS ONE IS UP-TO-DATE WITH THE SERVER

### CODE REVIEW

File: challenge_set_a_table/src/challenge_set_a_table.py [hunk 1]
```diff
@@ -40,11 +43,31 @@ def __init__(self, robot):
...
+                                                             radius=0.7),
+                                   transitions={'arrived': 'ANNOUNCEMENT',
+                                                'unreachable': 'ANNOUNCEMENT',
+                                                'goal_not_defined': 'ANNOUNCEMENT'})
             # Part I: Set a table
-            smach.StateMachine.add('ENTER_ROOM',  # Enter the room
-                                   states.Initialize(robot),
-                                   transitions={'initialized': 'ANNOUNCEMENT',
-                                                'abort': 'Aborted'})
+            # smach.StateMachine.add('ENTER_ROOM',  # Enter the room
```
Comment (jlunenburg) [inline challenge_set_a_table/src/challenge_set_a_table.py]: Remove this

File: challenge_spr/src/challenge_spr_states/riddle_game.py [hunk 2]
```diff
@@ -171,22 +173,48 @@ def answer_placement_location(action):
...
+    objects = [obj for obj in common_knowledge.objects if obj['name'] == entity]
+    if not objects:
+        return "I should count but I don't know that object %s" % entity
+
+    if objects[0]["category"] in common_knowledge.category_locations:
+        (found_location, area_name) = common_knowledge.get_object_category_location(objects[0]["category"])
+        if found_location == location:
+            return 'There is one %s in the %s' % (entity, location)
+    return 'There are no %s in the %s' % (entity, location)
+    #
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply challenge_spr/src/challenge_spr_states/riddle_game.py]: Cleanup in 9942c1dc77a35d4e4c0ab7271bffe5704058f9d8

File: challenge_set_a_table/src/challenge_set_a_table_states/fetch_command.py [hunk 3]
```diff
@@ -180,7 +180,7 @@ def _recite(self, choice):
                                                                    self.options[choice]["drink2"]), block=False)
         self.robot.speech.speak("To eat, you will get {} and {}".format(self.options[choice]["difficult1"],
                                                                         self.options[choice]["difficult2"]), block=False)
-        self.robot.speech.speak("I will serve this all on a {}".format(self.options[choice]["difficult3"]), block=True)
+        # self.robot.speech.speak("I will serve this all on a {}".format(self.options[choice]["difficult3"]), block=True)
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply challenge_set_a_table/src/challenge_set_a_table_states/fetch_command.py]: Fixed in 7219f1f196872af02c0da3dd0b237f66f0993cc3

File: robot_skills/src/robot_skills/robot_part.py [hunk 4]
```diff
@@ -176,7 +176,8 @@ def process_hardware_status(self, diagnostic_dict):
         diag = diagnostic_dict.get(self.__diagnostics_name, None)
 
         if not diag:
-            rospy.logwarn('no diagnostic msg received for the %s' % self.__diagnostics_name)
+            pass
+	    #rospy.logwarn('no diagnostic msg received for the %s' % self.__diagnostics_name)
```
Comment (Rayman) [inline robot_skills/src/robot_skills/robot_part.py]: Why is this warning removed?

Comment (LoyVanBeek) [reply robot_skills/src/robot_skills/robot_part.py]: Good catch. 

Because it was annoying, apparently. Needs a good looking after for Sergio, as that does miss some bodyparts that exist in code but not reality and thus things will complain at a high rate when diagnostics come in

Comment (MaxBaeten) [reply robot_skills/src/robot_skills/robot_part.py]: It was removed because it showed up with 100Hz frequency or something and everying else in your terminal becomes invisible at that point.

Comment (Timple) [reply robot_skills/src/robot_skills/robot_part.py]: Lets just make it a rospy.logwarnonce and be done with the issue on short term ;-)

Comment (jlunenburg) [reply robot_skills/src/robot_skills/robot_part.py]: Logging Once
New in Lunar

Since Lunar, rospy supports writing log messages only once after it is spawned. The output will print a message once by `rospy.log*_once(msg)`.

```
 while True:
    rospy.loginfo_once("This message will print only once")
```

Comment (Timple) [reply robot_skills/src/robot_skills/robot_part.py]: `rospy.loginfo_throttle(Inf, "This message will print every infinity seconds, so basically just once")`

:-D

Comment (jlunenburg) [reply robot_skills/src/robot_skills/robot_part.py]: Like this: 44eecf6bc5e2c6c4da7b23c26e75f53e6a420849 ??

File: challenge_spr/src/challenge_spr_states/riddle_game.py [hunk 5]
```diff
@@ -171,22 +177,41 @@ def answer_placement_location(action):
...
+            return 'There is one %s in the %s' % (entity, location)
+    return 'There are no %s in the %s' % (entity, location)
+
+
 def answer_find_objects(action):
     entity = action['entity']
     objects = [obj for obj in common_knowledge.objects if obj['name'] == entity]
     if len(objects) == 1:
         cat = objects[0]['category']
-        print cat
```
Comment (rokusottervanger) [inline challenge_spr/src/challenge_spr_states/riddle_game.py]: Remove plain prints. If you need to report to the user, use logging tools. If you don't, then don't print at all.

File: challenge_storing_groceries/src/challenge_storing_groceries/storing_groceries.py [hunk 6]
```diff
@@ -43,34 +44,38 @@ def __init__(self, robot):
                                                 'no_response': 'AWAIT_START'})
 
             cabinet = ds.EntityByIdDesignator(robot, id=CABINET)
-            room = ds.EntityByIdDesignator(robot, id=ROOM)
+            # room = ds.EntityByIdDesignator(robot, id=ROOM)
```
Comment (rokusottervanger) [inline challenge_storing_groceries/src/challenge_storing_groceries/storing_groceries.py]: If this room is not required, it doesn't need to be commented out, right?

File: robocup_knowledge/src/robocup_knowledge/environments/rgo2018/challenge_gpsr.py [hunk 7]
```diff
@@ -1,20 +1,18 @@
 from robocup_knowledge import knowledge_loader
 common = knowledge_loader.load_knowledge("common")
 
+names = ['josha', 'lars', 'janno']
+
```
Comment (rokusottervanger) [inline robocup_knowledge/src/robocup_knowledge/environments/rgo2018/challenge_gpsr.py]: This was a hack for the final. Should be removed.

File: challenge_set_a_table/src/challenge_set_a_table.py [hunk 8]
```diff
@@ -71,7 +94,7 @@ def __init__(self, robot):
                                                      grasp_designator1=grasp_designator1,
                                                      grasp_designator2=grasp_designator2,
                                                      grasp_designator3=grasp_designator3,
-                                                     grasp_furniture_id1=knowledge.cupboard,
+                                                     grasp_furniture_id1=knowledge.kitchen,
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply challenge_set_a_table/src/challenge_set_a_table.py]: @KevinDang74 can you comment on this and/or fix it?

File: robot_smach_states/src/robot_smach_states/human_interaction/learn_operator.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                    if self._robot.perception.learn_person(self._operator_name):
+                        rospy.loginfo("Succesfully detected you %i times" % (num_detections + 1))
+                        num_detections += 1
+                    elif (rospy.Time.now() - learn_person_start_time).to_sec() > self._learn_person_timeout:
+                        self._robot.speech.speak("Please stand in front of me and look at me")
+                        operator = None
+                        break
+            r.sleep()
+        rospy.loginfo("We have a new operator: %s" % operator.id)
+        self._robot.speech.speak("Gotcha! I will follow you!", block=False)
```
Comment (MatthijsBurgh) [inline robot_smach_states/src/robot_smach_states/human_interaction/learn_operator.py]: This state could be used in more cases than following, therefore I would prefer to replace this line with something generic.

File: challenge_storing_groceries/src/challenge_storing_groceries/storing_groceries.py [hunk 10]
```diff
@@ -43,34 +44,38 @@ def __init__(self, robot):
...
-                robot.ed.update_entity(id=table_id, frame_stamped=ENTITY_POSES[0][1])
+                # Update the world model by fitting the entities to the frame_stamped's given below.
+                robot.ed.update_entity(id=cabinet_id, frame_stamped=closest_workspace.place_entity_conf.pose_estimate)
+                robot.ed.update_entity(id=table_id, frame_stamped=closest_workspace.grasp_entity_conf.pose_estimate)
 
                 # Update designators
-                cabinet.id_ = ENTITY_POSES[0][2]
-                room.id_ = ENTITY_POSES[0][4]
+                cabinet.id_ = closest_workspace.place_entity_conf.entity_id
+                # room.id_ = ENTITY_POSES[0][4] # TODO: no usages besides this line and instantiation?
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply challenge_storing_groceries/src/challenge_storing_groceries/storing_groceries.py]: @LoyVanBeek please fix ASAP!!!

File: robot_smach_states/examples/guide-node [hunk 11]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/python
```
Comment (rokusottervanger) [inline robot_smach_states/examples/guide-node]: Is this file a good example of how the guide state machine should be used? Why is this here?

File: robot_skills/src/robot_skills/arms.py [hunk 12]
```diff
@@ -569,3 +569,199 @@ def _publish_marker(self, goal, color, ns = ""):
 
     def __repr__(self):
         return "Arm(side='{side}')".format(side=self.side)
+
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/arms.py]: @Jip238 did you work on this? Please fix it ASAP, it's blocking the merge!

(in principle a new feature should *always* be committed in a feature branch so that it can be reviewed separately)

File: robot_smach_states/examples/test-detect-faces [hunk 13]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/python
```
Comment (rokusottervanger) [inline robot_smach_states/examples/test-detect-faces]: I see this is a test of the DetectFace state, but it is in the 'examples' directory. Is it an example, or a test?

File: robot_skills/src/robot_skills/lights.py [hunk 14]
```diff
@@ -29,6 +29,7 @@ def close(self):
         pass
 
     def set_color(self, r, g, b, a=1.0):
+        # type: (object, object, object, object) -> object
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/lights.py]: Fixed in 0c3849b3c1f443a99a60c2cd0626089285c0dd62

File: robot_smach_states/src/robot_smach_states/navigation/guide.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+         * {EdEntityDesignator: str} mapping EdEntityDesignators to a string or designator resolving to a string,
+           representing the area, e.g., entity_designator_area_name_map[<EdEntity>] = 'in_front_of'.
+        :param follower: (EdEntityDesignator) (optional) designates the person (or robot) to guide
+        """
+        smach.StateMachine.__init__(self, outcomes=["succeeded", "failed", "preempted"])
+
+        nav_machine = create_navigation_state(robot, target_location)
+
+        # ToDo: update
+        loc_sentence = "Follow me if you want to make it out of here alive"
```
Comment (rokusottervanger) [inline robot_smach_states/src/robot_smach_states/navigation/guide.py]: I think there is a quote in the sound board that says "Follow me if you want to live". But I could be mistaken.

File: robot_skills/src/robot_skills/robot.py [hunk 16]
```diff
@@ -45,8 +45,15 @@ def __init__(self, robot_name="", wait_services=False):
         self.parts = dict()
         self.parts['base'] = base.Base(self.robot_name, self.tf_listener)
         self.parts['torso'] = torso.Torso(self.robot_name, self.tf_listener)
-        self.parts['leftArm'] = arms.Arm(self.robot_name, self.tf_listener, side="left")
-        self.parts['rightArm'] = arms.Arm(self.robot_name, self.tf_listener, side="right")
+
+        # ToDo: move this to subclasses
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/robot.py]: @Jip238 is this yours? Please address it ASAP!!!

File: robot_skills/src/robot_skills/world_model_ed.py [hunk 17]
```diff
@@ -256,16 +267,16 @@ def lock_entities(self, lock_ids, unlock_ids):
...
-        #if isinstance(center_point, PointStamped):
+    def get_closest_possible_person_entity(self, type="", center_point=VectorStamped(), radius=0, room=""):
+        # if isinstance(center_point, PointStamped):
         #    center_point = self._transform_center_point_to_map(center_point)
         # ToDo: check frame ids
 
         entities = self.get_entities(type="", center_point=center_point, radius=radius)
 
         # HACK
         # entities = [ e for e in entities if e.convex_hull and e.type == "" and 'possible_human' in e.flags ]
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/world_model_ed.py]: Updated in 1f0d46972e51b08ff55af3de36066fccda6f66cc

File: challenge_spr/src/challenge_spr_states/riddle_game.py [hunk 18]
```diff
@@ -171,22 +173,48 @@ def answer_placement_location(action):
...
+    locations = [loc for loc in locations if loc['room'] == location]
+    if not locations:
+        return 'There are no %s in the %s' % (entity, location)
+    elif len(locations) == 1:
+        return 'There is one %s in the %s' % (entity, location)
+    else:
+        return 'There are %d %s in the %s' % (len(locations), entity, location)
+
+
+def answer_count_objects_placement(action):
```
[the remark this answers is not in the document]
Comment (rokusottervanger) [reply challenge_spr/src/challenge_spr_states/riddle_game.py]: @SamAlexandrov , could you review, clean up, and add docstrings here?

File: robot_skills/src/robot_skills/world_model_ed.py [hunk 19]
```diff
@@ -413,3 +425,34 @@ def _publish_marker(self, center_point, radius):
         marker.color.r = 1
 
         self._marker_publisher.publish(marker)
+
+    def sync_furniture_poses(self):
+        """ Syncs the furniture poses of the world model to disc
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/world_model_ed.py]: Update in 88ba8ef1ba8634d1ecfbb17053d0596da7302017

File: robot_smach_states/src/robot_smach_states/human_interaction/detect_face.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
+# ROS
+import rospy
+import smach
+from std_msgs.msg import String
+from sensor_msgs.msg import Image
+
+
+class DetectFace(smach.State):
+    """
+    Detect face ToDo: expand docstring
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_smach_states/src/robot_smach_states/human_interaction/detect_face.py]: Update in 84e63633151ff0cb2685e7171f618a1b7873a8b3

File: robot_smach_states/src/robot_smach_states/human_interaction/find_person_in_room.py [hunk 21]
```diff
@@ -0,0 +1,247 @@
+#! /usr/bin/env python
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_smach_states/src/robot_smach_states/human_interaction/find_person_in_room.py]: Update in d629019c5abd445b329ce2bc164202db4060b758 
Shebang required for test (bottom of file), prefer to leave it like this

File: challenge_set_a_table/src/challenge_set_a_table.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (KevinDang74) [reply challenge_set_a_table/src/challenge_set_a_table.py]: Update in 43a7bb0845162ae3052535a77dab16dfdc149621

File: robot_skills/src/robot_skills/arms.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Jip238) [reply robot_skills/src/robot_skills/arms.py]: I don't recall working on this. Could you take a look at it?

File: robot_skills/src/robot_skills/robot.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Jip238) [reply robot_skills/src/robot_skills/robot.py]: Same, I don't recall working on this either. Could you check on this as well?

File: challenge_final/scripts/dirty_node.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply challenge_final/scripts/dirty_node.py]: I think it is too dirty. I will make a proposal about how we could keep the current functionality, but with more generic code and more functionality as well.

File: robot_skills/src/robot_skills/arms.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/arms.py]: Git blame: @Rayman 
I have cleaned up all the stuff which is duplicated.

File: robot_skills/src/robot_skills/robot.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/robot.py]: I moved all the fakearm stuff to sergio. The old implementation in the generic robot class is restored.

File: challenge_storing_groceries/src/challenge_storing_groceries/storing_groceries.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (LoyVanBeek) [reply challenge_storing_groceries/src/challenge_storing_groceries/storing_groceries.py]: Done in da1408d19890074cb37b6073ee83a0c0da57ab51

File: challenge_storing_groceries/src/challenge_storing_groceries/storing_groceries.py (hunk 6 above)
Comment (LoyVanBeek) [reply challenge_storing_groceries/src/challenge_storing_groceries/storing_groceries.py]: Done in 0fb7608e438c92a8c3522274d2880d9b837ef0a5

File: challenge_final/scripts/dirty_node.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Rayman) [reply challenge_final/scripts/dirty_node.py]: Fixed

File: challenge_restaurant/src/challenge_restaurant/restaurant.py [hunk 23]
```diff
@@ -60,7 +60,7 @@ def __init__(self, robot):
 
             smach.StateMachine.add('NAVIGATE_TO_CUSTOMER',
                                    states.NavigateToObserve(robot=robot, entity_designator=caller_designator,
-                                                            radius=0.85),
+                                                            radius=1.1),
```
[the remark this answers is not in the document]
Comment (LarsJanssenTUe) [reply challenge_restaurant/src/challenge_restaurant/restaurant.py]: This is done to increase robustness for SERGIO, since SERGIO is a lot bigger than AMIGO. Although I agree that 1.1 meter might be to far. I suggest to split in half way and make it 1m.

File: robot_skills/src/robot_skills/arms.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Rayman) [reply robot_skills/src/robot_skills/arms.py]: This is to get sergio working. All challenges except `robot.leftArm` to exist.

**Comment by Rayman** [comment]
Let's go then
